pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/mergestrategy/TimeBasedMergeStrategy.java

```diff
@@ -0,0 +1,133 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.controller.helix.core.minion.mergestrategy;
+
+import com.google.common.collect.Lists;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.concurrent.TimeUnit;
+import org.apache.pinot.common.metadata.segment.SegmentZKMetadata;
+import org.apache.pinot.core.common.MinionConstants;
+import org.apache.pinot.core.operator.transform.transformer.timeunit.CustomTimeUnitTransformer;
+import org.apache.pinot.spi.utils.TimeUtils;
+import org.joda.time.Interval;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Time based merge strategy.
+ *
+ * When the segments are correctly time aligned (e.g. each segment contains exactly 1 day's data), the time based merge
+ * strategy allows to merge segments based on the segment granularity.
+ *
+ * For example, the merge strategy will try to merge segments based on the daily bucket if the segment granularity is
+ * set to 'DAILY' (i.e. only segments from the same day can be considered to be merged).
+ *
+ * Default is 'NONE' where the strategy would try to merge everything into a single time bucket.
+ */
+public class TimeBasedMergeStrategy implements MergeStrategy {
+
+  public enum SegmentGranularity {
+    NONE, DAYS, WEEKS, MONTHS
+  }
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(TimeBasedMergeStrategy.class);
+
+  private SegmentGranularity _segmentGranularity;
+
+  public TimeBasedMergeStrategy(Map<String, String> taskConfigs) {
+    String segmentGranularityStr = taskConfigs.get(MinionConstants.MergeRollupTask.SEGMENT_GRANULARITY);
+    try {
+      _segmentGranularity = SegmentGranularity.valueOf(segmentGranularityStr);
+    } catch (Exception e) {
+      _segmentGranularity = SegmentGranularity.NONE;
+    }
+    LOGGER.info("Time based merge strategy is initialized with the segment granularity : {}", _segmentGranularity);
+  }
+
+  @Override
+  public List<List<SegmentZKMetadata>> generateMergeTaskCandidates(List<SegmentZKMetadata> segmentsToMerge,
```

Comment:
Per the design doc, each time we only merge segments for the earliest one day/week/month, where segments that cross the time boundary are split into 2 parts. Am I missing something here?